lib/datasource.js

```diff
@@ -538,24 +524,63 @@ DataSource.prototype.defineRelations = function(modelClass, relations) {
       if ((targetModel && !isModelDataSourceAttached(targetModel)) ||
           (throughModel && !isModelDataSourceAttached(throughModel))) {
         // Create a listener to defer the relation set up
-        createListener(rn, r, targetModel, throughModel);
+        deferRelationSetup(relationName, r, targetModel, throughModel);
       } else {
         // The target model is resolved
         var params = traverse(r).clone();
-        params.as = rn;
-        params.model = polymorphicName || targetModel;
+        params.as = relationName;
+        params.model = targetModel;
         if (throughModel) {
           params.through = throughModel;
         }
-        modelClass[r.type].call(modelClass, rn, params);
+        modelClass[r.type].call(modelClass, relationName, params);
       }
     });
   }
 };
 
+function validateRelation(relationName, relation) {
+  var rn = relationName;
+  var r = relation;
+
+  assert(DataSource.relationTypes.indexOf(r.type) !== -1, 'Invalid relation type: ' + r.type);
+  assert(isValidRelationName(rn), 'Invalid relation name: ' + rn);
+
+  // non polymorphic belongsTo relations should have `model` defined
+  if (!r.polymorphic && r.type === 'belongsTo' && !r.model) {
+    throw new Error(g.f('%s relation: %s requires param `model`', r.type, rn));
+  }
+  // polymorphic belongsTo relations should not have `model` defined
+  if (r.polymorphic && r.type === 'belongsTo' && r.model) {
+    throw new Error(g.f('{{polymorphic}} %s relation: %s does not expect param `model`', r.type, rn));
+  }
+  // polymorphic relations other than belongsTo should have `model` defined
+  if (r.polymorphic && r.type !== 'belongsTo' && !r.model) {
+    throw new Error(g.f('{{polymorphic}} %s relation: %s requires param `model`', r.type, rn));
+  }
+  // polymorphic belongsTo relations should not have `polymorphic.as` defined
+  if (r.polymorphic && r.type === 'belongsTo' && r.polymorphic.as) {
+    throw new Error(g.f('Unexpected param `polymorphic.as` in {{polymorphic}} %s ' +
+      'relation: %s', r.type, rn));
+  }
+  // polymorphic belongsTo relations should not have `polymorphic` of type 'string'
+  if (r.type === 'belongsTo' && typeof r.polymorphic === 'string') {
+    throw new Error(g.f('Unexpected type \'string\' for param `polymorphic` in ' +
+      '{{polymorphic}} %s relation: %s', r.type, rn));
+  }
+  if (r.polymorphic && r.polymorphic.foreignKey && !r.polymorphic.discriminator) {
+    throw new Error(g.f('{{polymorphic}} %s relation: %s requires param `polymorphic.discriminator` ' +
+      'when param `polymorphic.foreignKey` is provided', r.type, rn));
+  }
+  if (r.polymorphic && r.polymorphic.discriminator && !r.polymorphic.foreignKey) {
+    throw new Error(g.f('{{polymorphic}} %s relation: %s requires param `polymorphic.foreignKey` ' +
+      'when param `polymorphic.discriminator` is provided', r.type, rn));
+  }
+  // TODO: warn on extraneous `polymorphic.as` keyword
+}
```

Comment:
arf... i'd prefer not put a reference to `polymorphic.as` : the existing behavior always favor `discriminator` and `foreignKey` over `as` so your statement is not correct.